PLU-275: [BTN-MIGRATION-2] Postman SMS app

ogp-weeloong opened the pull request:
## Problem
We need to migrate our Twilio users to Postman SMS. For that to happen, we need to expose the Postman SMS functionality to them first.

## Solution
We will create a new "SMS by Postman" app. For now, we will only support sending SMSes via Plumber's API key (similar to our "Email by Postman" approach).

When we get Postman to code up a "get SMS templates" API, we'll allow users to add their own connections and choose their templates. 

<img width="500" alt="Screenshot 2024-06-03 at 2 39 00 PM" src="https://github.com/opengovsg/plumber/assets/135598754/3b4dc2ba-06d8-4618-8a7c-a5db33dffba3">

## Tests
- New unit test
- Check that SMSes are sent

## Deploy Notes
**New environment variables**:
- `POSTMAN_SMS_API_BASE_URL`: To allow switching between postman test and prod env on Plumber staging/uat vs prod 
- `POSTMAN_SMS_DEFAULT_CAMPAIGN_ID`: Campaign ID for Plumber's postman SMS campaign
- `POSTMAN_SMS_DEFAULT_API_KEY`: API key for Plumber's Postman SMS campaign
- `POSTMAN_SMS_QPS_LIMIT_PER_CAMPAIGN`: Postman's campaign rate limit

ogp-weeloong commented:
* **#607** <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/607?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>: 2  dependent PRs ([#616](https://github.com/opengovsg/plumber/pull/616) <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/616?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>, [#625](https://github.com/opengovsg/plumber/pull/625) <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/625?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>)
* **#606** <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/606?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#605** <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/605?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `develop-v2`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @ogp-weeloong and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

pregnantboy commented on packages/backend/src/apps/postman-sms/actions/send-sms/schema.ts:
```diff
@@ -0,0 +1,22 @@
+
+export const fieldSchema = z.object({
+  recipient: telephoneNumberSchema,
+  message: z.string().trim().min(1, { message: 'Provide a non-empty message' }),
```
can add max length here

pregnantboy replied:
ok maybe not since the 1000 characters includes the standard header and footer

ogp-weeloong replied:
actually can add lol

pregnantboy commented on packages/backend/src/config/app-env-vars/postman-sms.ts:
```diff
@@ -0,0 +1,27 @@
+}
+
+if (
+  isNaN(postmanSmsConfig.qpsLimitPerCampaign) ||
```
nit: !isInteger

ogp-weeloong replied:
good catch

pregnantboy commented on packages/backend/src/apps/postman-sms/queue.ts:
```diff
@@ -0,0 +1,29 @@
+import type { IAppQueue } from '@plumber/types'
```
why for m365 u create the file under queue/index.ts but here u create as queue.ts

ogp-weeloong replied:
hey pi can liao

actually i create under queue bah for consistency

pregnantboy commented on packages/backend/.env-example:
```diff
@@ -28,6 +28,11 @@ M365_SG_GOVT_CLIENT_PRIVATE_KEY=...
+POSTMAN_SMS_DEFAULT_CAMPAIGN_ID=...
+POSTMAN_SMS_DEFAULT_API_KEY=...
+POSTMAN_SMS_QPS_LIMIT_PER_CAMPAIGN=3
+
```
can add into 1pw?

m0nggh replied:
+1 thkss

pregnantboy commented on packages/backend/src/apps/postman-sms/actions/send-sms/index.ts:
```diff
@@ -0,0 +1,95 @@
+    },
+    {
+      label: 'Message',
+      description: 'Max 1000 characters',
```
... including default headers and footers

ogp-weeloong replied:
[excluding](https://postman-v2.guides.gov.sg/faq/postman-v2-api-faq/campaign-related-inquiries)

>  The total message length (excluding header and footer) shall not exceed 1000 characters.

pregnantboy commented on ecs/env.json:
```diff
@@ -197,6 +197,22 @@
+    },
+    {
+      "name": "POSTMAN_SMS_API_BASE_URL",
+      "valueFrom": "plumber-<ENVIRONMENT>-postman-sms-api-base-url"
```
in future, can probably be stored with the connection like paysg

ogp-weeloong replied:
actually i might as well change it to this approach now

ogp-weeloong replied:
done, lmk what you think

pregnantboy commented on packages/backend/src/apps/postman-sms/queue.ts:
```diff
@@ -0,0 +1,29 @@
+      duration: 1000,
+    },
+  },
+  isQueueDelayable: false,
```
can add 429 handler to delay by group
https://postman-v2.guides.gov.sg/general-notes-for-api-users/rate-limits

pregnantboy reviewed (approved):
tested. lgtm with minor nits.

pregnantboy reviewed (changes requested):
429 handler, the rest lgtm

pregnantboy commented on packages/backend/src/apps/postman-sms/actions/send-sms/index.ts:
```diff
@@ -0,0 +1,95 @@
+          },
+        },
+      )
+      const message = postmanMessageSchema.parse(response.data)
```
this is a bit dangerous since the SMS should have successfully sent out at this stage. The only reason this could throw an error is if postman changes their response, so instead of failing the step, we should log it instead. can try using safeParse instead to prevent termination of step

ogp-weeloong replied:
For my future reference: the original thinking is that if postman changes their response, then the variables the users configured in subsequent steps will be invalid; hence we should error out instead of silently failing. This is at the tradeoff of re-sending the SMS if user retries the step

However, as discussed, its a non-goal to handle downstream API changes if they do not inform us first (since this should not happen). So will switch to logging and letting this step succeed.

Plus, most likely people should not have steps _after_ sending SMSes.

pregnantboy commented on packages/backend/src/apps/postman-sms/actions/send-sms/index.ts:
```diff
@@ -0,0 +1,95 @@
+const action = {
+  name: 'Send SMS',
+  key: 'sendSms',
+  description: 'Sends an SMS; display name is set to plumber.gov.sg',
```
```Sends an SMS under Gov.SG sender ID```
in future when credentials are added, it'll no longer be under plumber.gov.sg

ogp-weeloong replied:
done

ogp-weeloong commented on packages/backend/src/apps/postman-sms/actions/send-sms/index.ts:
```diff
@@ -0,0 +1,119 @@
+      // We'll allow the step to succeed as SMS has already been sent out, but
+      // we don't expose any of the response data because we have no clue what
+      // it contains now. Returning empty is than returning gibberish, because
+      // the latter leads the user into a false sense of correctness.
```
@pregnantboy lmk what you think of this

ogp-weeloong commented:
v8 changes:
* Removed base URL env var
* Validate max message length
* Allow step to succeed even if postman's response changed
* Handle 429 from Postman via a group-delay retry
* Fixed phone number regex not rejecting invalid numbers as long as there was at least 1 valid character
* Added more tests

pregnantboy commented on packages/backend/src/apps/postman-sms/actions/send-sms/index.ts:
```diff
@@ -0,0 +1,119 @@
+          $.step.position,
+          $.app.name,
+        )
+      } else {
```
nit: unnecessary else

ogp-weeloong replied:
nice, fixed

pregnantboy commented on packages/backend/src/apps/postman-sms/actions/send-sms/index.ts:
```diff
@@ -0,0 +1,119 @@
+        eventName: 'sendSms',
+      })
+      $.setActionItem({
+        raw: {},
```
lgtm, u wanna add a createdAt as current time ?

pregnantboy replied:
with id blank

ogp-weeloong replied:
yep, done

ogp-weeloong commented:
### Merge activity

* **Jun 14, 5:49 AM EDT**: @ogp-weeloong started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/opengovsg/plumber/606).
* **Jun 14, 5:50 AM EDT**: @ogp-weeloong merged this pull request with [Graphite](https://app.graphite.dev/github/pr/opengovsg/plumber/606).
